build_wheels.py

```diff
@@ -0,0 +1,227 @@
+        idf_repo_branches (list): IDF fetched branches to select from
+
+    Returns:
+        list: IDF branches to download requirements for
+    """
+    idf_branches = []
+    for branch in idf_repo_branches:
+        try:
+            idf_major, idf_minor = re.match(r'release/v(\d+)\.(\d+)', branch).groups()  # type: ignore
+            if (int(idf_major), int(idf_minor)) >= (constants.MIN_IDF_MAJOR_VERSION, constants.MIN_IDF_MINOR_VERSION):
+                idf_branches.append(branch)
+        except AttributeError:
+            continue
+    idf_branches.append('master')
+    return idf_branches
+
+
+# Constraints files versions list
+def get_constraints_versions(idf_branches: list) -> list:
+    """From desired branches passed in get constraints files versions list
+    - when branch is not 'release' (without version) it is supposed to be 'master'
+    and auto version mechanism is applied if not specified in constants.py not to
+
+    Args:
+        idf_branches (list): IDF branches to download requirements for
+
+    Returns:
+        list: IDF constraints files versions to download requirements for in right order to the branches list
+    """
+    idf_constraints = []
+
+    def get_master_constraints(idf_constraints: list):
+        """Automatically get master branch constraints file version by incrementing last release minor field by 1
+
+        Args:
+            idf_constraints (list): IDF constraints files versions to download requirements for
+        """
+        idf_last_major, idf_last_minor = re.match(r'v(\d+)\.(\d+)', idf_constraints[-1]).groups()  # type: ignore
+        idf_master_minor = str(int(idf_last_minor)+1)
+        if requests.get(f'{idf_constraints_url}v{idf_last_major}.{idf_master_minor}.txt').status_code == 200:
+            idf_constraints.append(f'v{idf_last_major}.{idf_master_minor}')
+        else:
+            idf_constraints.append(idf_constraints[-1])
+
+    for branch in idf_branches:
+        try:
+            splitted_branch = branch.split('release/')
+            idf_constraints.append(splitted_branch[1])
+        except IndexError:
+            if constants.AUTO_VERSION_CONSTRAINS_FILE:
+                get_master_constraints(idf_constraints)
+            else:
+                idf_constraints.append(constants.MASTER_CONSTRAINS_FILE)
+    return idf_constraints
+
+
+# --- Download all requirements from all the branches requirements and constraints files --- #
+def assemble_requirements(idf_branches: list, idf_constraints: list) -> set:
+    """Assemble IDF requirements into set to prevent duplicates
+
+    Args:
+        idf_branches (list): IDF branches to download requirements for
+        idf_constraints (list): IDF constraints files versions to download requirements for
+
+    Returns:
+        set: IDF requirements
+    """
+    idf_resources_url = 'https://raw.githubusercontent.com/espressif/esp-idf/'
+    requirements_txt: list = []
+
+    def download_branch_requirements() -> list:
```

Comment:
I would prefer not nested functions.

For example, if `def download_branch_requirements() -> list:` is a helper function, you can mark it as such with an underscore before the function name and move it up a level:

```py
def _download_branch_requirements(branch: str, idf_requirements_json: dict) -> list:
    """Download requirements files for all groups specified in IDF requirements.JSON."""
    print(f'---------- ESP-IDF BRANCH {branch} ----------')
    requirements_txt = []
    for feature in idf_requirements_json['features']:
        res = requests.get(f"{IDF_RESOURCES_URL}{branch}/{feature['requirement_path']}")
        if check_response(res, f"Failed to download feature (requirement group) '{feature['name']}'"):
            requirements_txt.extend(res.text.splitlines())
            print(f"Added ESP-IDF {feature['name']} requirements")
    return requirements_txt
```

The same for `_download_branch_constraints` and `_add_into_requirements`.